prune.go

```diff
@@ -129,122 +135,194 @@ func txnsPruneC(txnsName string) string {
+		return fmt.Errorf("reading collection names: %v", err)
 	}
-	if err := iter.Close(); err != nil {
-		return fmt.Errorf("failed to read all txns: %v", err)
+	var collNames []string
+	for _, doc := range collNameDocs {
+		if len(doc) == 1 {
+			if collName, ok := doc[0].Value.(string); ok {
+				collNames = append(collNames, collName)
+			}
+		}
 	}
-	logger.Debugf("found %d completed transactions across %d collections",
-		len(txnIds), len(collNames))
+
+	logger.Debugf("%d collections with txns discovered", len(collNames))
 
 	// Transactions may also be referenced in the stash.
-	collNames["txns.stash"] = present
+	collNames = append(collNames, "txns.stash")
 
-	// Now remove the txn ids referenced by all documents in all
-	// txn using collections from the set of known txn ids.
+	// Now remove the txn ids referenced by any document in any
+	// txn-using collection from the set of known txn ids.
 	//
 	// Working the other way - starting with the set of txns
 	// referenced by documents and then removing any not in that set
 	// from the txns collection - is unsafe as it will result in the
-	// removal of transactions run while pruning executes.
-	//
-	for collName := range collNames {
-		logger.Tracef("checking %s for transaction references", collName)
+	// removal of transactions created during the pruning process.
+	t := newSimpleTimer(logInterval)
+	remover := newBulkRemover(workingSet)
+	defer remover.close()
+	for _, collName := range collNames {
+		logger.Tracef("checking %s for txn references", collName)
 		coll := db.C(collName)
 		var tDoc struct {
 			Queue []string `bson:"txn-queue"`
 		}
-		iter := coll.Find(nil).Select(bson.M{"txn-queue": 1}).Iter()
+		query := coll.Find(nil).Select(bson.M{"txn-queue": 1})
+		query.Batch(batchBytes)
+		iter := query.Iter()
 		for iter.Next(&tDoc) {
 			for _, token := range tDoc.Queue {
-				delete(txnIds, txnTokenToId(token))
+				if err := remover.remove(txnTokenToId(token)); err != nil {
+					return fmt.Errorf("handling completed txns: %v", err)
+				}
+				if t.isAfter() {
+					logger.Debugf("%d referenced txns found so far", remover.removed)
+				}
 			}
 		}
 		if err := iter.Close(); err != nil {
 			return fmt.Errorf("failed to read docs: %v", err)
 		}
 	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("handling completed txns: %v", err)
+	}
+	logger.Debugf("%d txns are still referenced and will be kept", remover.removed)
 
-	// Remove the unreferenced transactions.
-	logger.Debugf("%d transactions to remove", len(txnIds))
-	if err := bulkRemoveTxns(txns, txnIds); err != nil {
-		return fmt.Errorf("txn removal failed: %v", err)
+	// Remove the no-longer-referenced transactions from the txns collection.
+	t = newSimpleTimer(logInterval)
+	remover = newBulkRemover(txns)
+	defer remover.close()
+	query := workingSet.Find(nil).Batch(batchBytes)
+	iter := query.Iter()
+	var doc bson.D
+	for iter.Next(&doc) {
+		if err := remover.remove(doc[0].Value); err != nil {
```

Comment:
I'm curious why before you had to do the "if len() == 1" check, but here you don't